ARCHITECTURE.md

```diff
@@ -0,0 +1,30 @@
+# Architecture of Trust-DNS
+
+The Trust-DNS libraries are built from the ground up to be asynchronous. This project grew from first using non-blocking IO interfaces (before Futures 0.1 or async/await had landed in Rust). There are some artifacts of this history sporadically left around the project. Please feel free to submit PRs that clean up areas that still have hand-written Futures based state-machines. Additionally, much of the project was written by @bluejekyll while he learned the Rust language–this means that there may be patterns or missing common implementations in places where he didn't know better. Feel free to clean that up if you feel open to submitting a PR.
+
+## Layout
+
+Most of the project is in the form of crates. The crates are all individually published to crates.io.
+
+The project is made of four central crates:
```

Comment:
I'm not sure the central/runtime/abstractions division here is useful. I'd probably start with the high-level crates, resolver, server (and include async-std-resolver in this list as well).

The rest of the crates are then "implementation detail" crates. I do think it's useful to call out the crates that provide different implementations based on other libraries, though "for different abstractions" doesn't really convey the underlying idea to me.

Personally I feel the current number of crates is too high, and it would make more sense to use a different abstraction layer (modules + features) for the runtime/TLS support variants (as such, I'm not sure what problem the introduction of a separate tokio-resolver crate will solve).

Reply:
This is good feedback. The reason I decided to go to separate crates for so much of this was to ensure feature separation, though in the end maybe it doesn't matter. I'd love to discuss this more though.

Reply:
In Askama, the way I have integration crates is that integration crates expose some additional integration-specific interface on top of my core functionality. This is useful because each framework moves at a different pace, different web frameworks may have conflicting dependency requirement versions (*ring* has been an issue for that, in particular). However, in trust-dns the integration crates are depended upon by the "core" crates (proto, resolver), so this advantage really doesn't exist. The other use case for integration crates is usually integrations spanning administrative boundaries (you might be familiar with Conway's law), where it's useful to have an integration that can keep up with both sides' changes and is administered separately. That also isn't relevant here: you/we maintain trust-dns and all of the integrations in this repository.

So, the two routes that I think could make sense would either be (a) expose some traits in core crates (proto/resolver/etc) that integrations can implement and then remove all the dependencies from core crates, or (b) just pull everything into the core crates. I think we're broadly abstracting over two different things: async runtimes and TLS implementations. For the former, I think it makes sense to just keep that dependency in the core crate(s), since these move pretty slowly anyway (so maybe merge the functionality from async-std-resolver into trust-dns-resolver). For TLS implementations, there is maybe more variety and also more movement, so it could still make sense to keep those separate. In that case I think we should come up with a well-defined trait for `SecureTransport` (to be bikeshedded) that can advertise support for DoT/DoH. A resolver would then abstract over a `Runtime` and a `SecureTransport`. We could default to tokio + rustls and provide an alternative implementation for tokio + native-tls and/or async-std + rustls and/or async-std + native-tls.

If we have well-defined traits, we could also as a project chose to only maintain a tokio + rustls stack and leave other integrations to others.

Reply:
Maybe we should open an issue for this conversation. The original motivation for proto/server/client/resolver was keeping code reviews simple for the library use cases, i.e. folks who want to use resolver only need to review resolver & proto crates.

But, the server/client/resolver crates all need the various TLS implementations, so that's why they are external to those crates, but they could be pulled into the proto crate, I think that would make a degree of sense based on your reasoning above.

Creating better traits that represent these abstractions would of course be great, we're close with DnsHandle which is used by all the stream implementations...